File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> subSet = new ArrayList<>();
        helper(candidates, target, subSet, new ArrayList<>(), 0);
        return subSet;
    }

    public void helper(int[] candidates, int target, List<List<Integer>> subSet, List<Integer> temp, int start) {
        if (target==0) {
            subSet.add(new ArrayList<>(temp));
            // return;
        }
        for (int i = start; i < candidates.length;i++) {
            // if(temp!=target){
                if(candidates[i]>target){
                continue;
                }
            temp.add(candidates[i]);
            helper(candidates, target-candidates[i], subSet, temp, i);
            // }
            temp.remove(temp.size() - 1);
        }
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = []
        def helper(i,arr,target):
            # if i == len(candidates):
            if 0 == target:
                ans.append(arr[:])
                return
            if i == len(candidates) or target<0:
                return
            arr.append(candidates[i])
            helper(i,arr,target-candidates[i])
            arr.pop()
            helper(i+1,arr,target)
        helper(0,[],target)
        return ans
            
            
        